pallets/support/src/lib.rs

```diff
@@ -0,0 +1,165 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::too_many_arguments)]
+
+use crate::types::*;
+use frame_support::sp_runtime::app_crypto::sp_core;
+use frame_support::sp_runtime::{BoundedVec, DispatchError, DispatchResult};
+use frame_system::pallet_prelude::BlockNumberFor;
+use sp_core::ConstU32;
+use sp_std::vec::Vec;
+
+#[cfg(test)]
+mod tests;
+
+pub mod types;
+
+// Re-export pallet items so that they can be accessed from the crate namespace.
+pub use pallet::*;
+
+pub const MAX_STACK_SIZE: u32 = 16;
+
+const LOG_TARGET: &str = "runtime::amm-support";
+
+type ExecutionIdStack = BoundedVec<ExecutionType, ConstU32<MAX_STACK_SIZE>>;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	}
+
+	#[pallet::storage]
+	/// Next available incremental ID
+	#[pallet::getter(fn incremental_id)]
+	pub(super) type IncrementalId<T: Config> = StorageValue<_, IncrementalIdType, ValueQuery>;
+
+	#[pallet::storage]
+	/// Execution context to figure out where the trade is originated from
+	#[pallet::getter(fn execution_context)]
+	pub(super) type ExecutionContext<T: Config> = StorageValue<_, ExecutionIdStack, ValueQuery>;
+
+	#[pallet::error]
+	pub enum Error<T> {}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Trade executed.
+		Swapped {
+			swapper: T::AccountId,
+			filler: T::AccountId,
+			filler_type: Filler,
+			operation: TradeOperation,
+			inputs: Vec<Asset>,
+			outputs: Vec<Asset>,
+			fees: Vec<Fee<T::AccountId>>,
+			operation_stack: Vec<ExecutionType>,
+		},
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
+			ExecutionContext::<T>::kill();
+			
+			T::DbWeight::get().reads_writes(1, 1)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+}
+
+impl<T: Config> Pallet<T> {
+	pub fn deposit_trade_event(
+		swapper: T::AccountId,
+		filler: T::AccountId,
+		filler_type: Filler,
+		operation: TradeOperation,
+		inputs: Vec<Asset>,
+		outputs: Vec<Asset>,
+		fees: Vec<Fee<T::AccountId>>,
+	) {
+		let operation_stack = Self::get_context();
+		Self::deposit_event(Event::<T>::Swapped {
+			swapper,
+			filler,
+			filler_type,
+			operation,
+			inputs,
+			outputs,
+			fees,
+			operation_stack,
+		});
+	}
+
+	pub fn add_to_context<F>(execution_type: F) -> Result<IncrementalIdType, DispatchError>
```

Comment:
We want to have these methods infallible so I would change the return type to just `IncrementalIdType`, if possible. But I'm not sure when `try_mutate` can fail. This is just an idea.